src/transport/mod.rs

```diff
@@ -120,13 +131,179 @@ impl AsyncWrite for JetTransport {
 }
 
 pub trait JetStream: Stream {
-    fn shutdown(&mut self) -> std::io::Result<()>;
-    fn peer_addr(&self) -> Option<SocketAddr>;
     fn nb_bytes_read(&self) -> u64;
     fn set_packet_interceptor(&mut self, interceptor: Box<dyn PacketInterceptor>);
 }
 
 pub trait JetSink: Sink {
-    fn shutdown(&mut self) -> std::io::Result<()>;
     fn nb_bytes_written(&self) -> u64;
 }
+
+struct JetStreamImpl<T: AsyncRead> {
+    stream: ReadHalf<T>,
+    nb_bytes_read: Arc<AtomicU64>,
+    packet_interceptor: Option<Box<dyn PacketInterceptor>>,
+    peer_addr: Option<SocketAddr>,
+    peer_addr_str: String,
+    buffer: BipBufferWriter,
+}
+
+impl<T: AsyncRead> JetStreamImpl<T> {
+    fn new(
+        stream: ReadHalf<T>,
+        nb_bytes_read: Arc<AtomicU64>,
+        peer_addr: Option<SocketAddr>,
+        buffer: BipBufferWriter,
+    ) -> Self {
+        Self {
+            stream,
+            nb_bytes_read,
+            packet_interceptor: None,
+            peer_addr,
+            peer_addr_str: peer_addr.clone().map_or("Unknown".to_string(), |addr| addr.to_string()),
+            buffer,
+        }
+    }
+}
+
+impl<T: AsyncRead> Stream for JetStreamImpl<T> {
+    type Item = usize;
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        let peer_addr = &self.peer_addr_str;
+        let mut written = 0;
+        loop {
+            if let Some(mut reservation) = self.buffer.reserve(PART_LEN) {
```

Comment:
BipBuffer usage: try to reserve the buffer. If it's Some, then we will read to the buffer, truncate and send for the Sink